bootstrap/src/u128.ml

```diff
@@ -0,0 +1,1281 @@
+
+  let bit_not t =
+    {hi=U64.bit_not t.hi; lo=U64.bit_not t.lo}
+
+  let bit_sl = u128_bit_sl
+
+  let bit_usr = u128_bit_usr
+
+  let bit_ssr t i =
+    let i = i % 128 in
+    let hi = begin
+      if Rudiments_int.(i >= 64) then Int64.shift_right t.hi (i - 64)
+      else if Rudiments_int.(i > 0) then Int64.shift_right t.hi i
+      else t.hi
+    end in
+    let lo = begin
+      if Rudiments_int.(i >= 64) then Int64.shift_right t.hi (i - 64)
+      else if Rudiments_int.(i > 0) then
+        Int64.logor (Int64.shift_left t.hi (64 - i))
+          (Int64.shift_right_logical t.lo i)
+      else t.lo
+    end in
+    {hi; lo}
+
+  let ( + ) t0 t1 =
+    let lo = U64.(t0.lo + t1.lo) in
+    let carry = if U64.(lo < t0.lo) then U64.one else U64.zero in
+    let hi = U64.(t0.hi + t1.hi + carry) in
+    {hi; lo}
+
+  let ( - ) t0 t1 =
+    let borrow = if U64.(t0.lo < t1.lo) then U64.one else U64.zero in
+    let hi = U64.(t0.hi - t1.hi - borrow) in
+    let lo = U64.(t0.lo - t1.lo) in
+    {hi; lo}
+
+  let ( * ) t0 t1 =
+    (* Decompose inputs into arrays of 32-bit half-words, then use the standard
+       paper method of multi-digit multiplication, but in base 2^32.  The full
+       result requires m + n digits, where m and n are the number of input
+       digits in the muliplier and multiplicand.  For this function,
+       ndigits=m=n=4, and we only calculate/preserve the lowest ndigits
+       digits.
+
+       The digit arrays are encoded as (u32 array), which assures that only
+       significant bits are stored.  The intermediate computations use 64-bit
+       math so that two digits fit. *)
+    let hi32 x = U64.bit_usr x 32 in
+    let lo32 x = U64.(bit_and x (of_usize 0xffff_ffff)) in
+    let digits32 x = hi32 x, lo32 x in
+    let get arr i = U32.to_u64 (Caml.Array.get arr i) in
+    let set arr i x = Caml.Array.set arr i (U32.of_u64 x) in
+    let to_arr u = [|U32.of_u64 (lo32 u.lo); U32.of_u64 (hi32 u.lo);
+      U32.of_u64 (lo32 u.hi); U32.of_u64 (hi32 u.hi);|] in
+    let of_arr arr = begin
+      let hi = U64.(bit_or (bit_sl (get arr 3) 32) (get arr 2)) in
+      let lo = U64.(bit_or (bit_sl (get arr 1) 32) (get arr 0)) in
+      {hi; lo}
+    end in
+    let t0_arr = to_arr t0 in
+    let t1_arr = to_arr t1 in
+    let product = to_arr zero in
+    let ndigits = 4 in
+    let rec fn_i i j x = begin
+      match Rudiments_int.(i + j = ndigits) with
+      | true -> ()
+      | false -> begin
+          let x', d = digits32 (U64.((get t0_arr i) * (get t1_arr j)
+            + (get product Rudiments_int.(i + j)) + x)) in
+          set product Rudiments_int.(i + j) d;
+          fn_i (Rudiments_int.succ i) j x'
+        end
+    end in
+    let rec fn_j j = begin
+      match Rudiments_int.(j = ndigits) with
+      | true -> ()
+      | false -> begin
+          fn_i 0 j U64.zero;
+          fn_j (Rudiments_int.succ j)
+        end
+    end in
+    fn_j 0;
+    of_arr product
+
+  let of_string s =
```

Comment:
There was nothing in OCaml to piggyback this on, so here we have a state machine for lexing `u128`.  We'll need something similar for the compiler, at which point we can hopefully unify the implementations.